File: python/src/benchmarks/ycsb/generator.py
import math
import random
from typing import Any, Callable, Optional, Union

# Standard Zipfian skew parameter (0.99) defined in the YCSB specification.
DEFAULT_ZIPFIAN_CONSTANT = 0.99
DEFAULT_ITEM_COUNT = 10000000000
ZETAN_1K = 7.728953217284738
ZETAN_10B = 26.46902820178302
FNV_OFFSET_BASIS_64 = 0xCBF29CE484222325
FNV_PRIME_64 = 0x100000001B3


class ZipfianGenerator:
    """
    Generates random integers following a Zipfian distribution over [min_val, max_val].
    """

    def __init__(
        self,
        min_val: int,
        max_val: int,
        zipfian_constant: float = DEFAULT_ZIPFIAN_CONSTANT,
        zetan: Optional[float] = None,
    ):
        items = max_val - min_val + 1
        if items < 1:
            items = 1
        if zetan is None:
            zetan = ZipfianGenerator.compute_zeta(items, zipfian_constant)

        zeta2theta = 1.0 + math.pow(0.5, zipfian_constant)
        alpha = 1.0 / (1.0 - zipfian_constant)
        denom = 1.0 - zeta2theta / zetan
        if abs(denom) < 1e-12:
            eta = 1.0
        else:
            eta = (1.0 - math.pow(2.0 / float(items), 1.0 - zipfian_constant)) / denom

        self.items = items
        self.base = min_val
        self.zipfian_constant = zipfian_constant
        self.alpha = alpha
        self.zetan = zetan
        self.eta = eta
        self.zeta2theta = zeta2theta

    @staticmethod
    def compute_zeta(items: int, zipfian_constant: float) -> float:
        """
        Computes or approximates zeta(n, theta) = sum_{i=1}^n 1 / (i^theta).
        Uses exact summation up to 1000 items and Euler-Maclaurin integration for n > 1000.
        """
        if items <= 0:
            return 0.0

        if abs(zipfian_constant - DEFAULT_ZIPFIAN_CONSTANT) < 1e-9:
            if items == 1000:
                return ZETAN_1K
            if items == DEFAULT_ITEM_COUNT:
                return ZETAN_10B
            if items > 1000:
                return ZETAN_1K + (
                    math.pow(float(items), 1.0 - zipfian_constant)
                    - math.pow(1000.0, 1.0 - zipfian_constant)
                ) / (1.0 - zipfian_constant)
            return sum(
                1.0 / math.pow(float(i), zipfian_constant) for i in range(1, items + 1)
            )

        n0 = min(items, 1000)
        sum_zeta = sum(
            1.0 / math.pow(float(i), zipfian_constant) for i in range(1, n0 + 1)
        )
        if items > n0:
            sum_zeta += (
                math.pow(float(items), 1.0 - zipfian_constant)
                - math.pow(float(n0), 1.0 - zipfian_constant)
            ) / (1.0 - zipfian_constant)
        return sum_zeta

    def next_value(self) -> int:
        """Generates the next integer in the Zipfian distribution."""
        u = random.random()
        uz = u * self.zetan

        if uz < 1.0:
            return self.base

        if uz < self.zeta2theta:
            return self.base + 1

        val = int(
            float(self.items) * math.pow(self.eta * u - self.eta + 1.0, self.alpha)
        )
        val = max(0, min(self.items - 1, val))
        return self.base + val


class ScrambledZipfianGenerator:
    """
    Generates scrambled Zipfian distributed integers over [min_val, max_val].
    Scrambles the sequence across the 10-billion item space by hashing each Zipfian item with 64-bit FNV-1a.
    """

    def __init__(
        self,
        min_val: int,
        max_val: int,
        zipfian_constant: float = DEFAULT_ZIPFIAN_CONSTANT,
    ):
        self.min_val = min_val
        self.max_val = max_val
        self.item_count = max_val - min_val + 1
        if self.item_count < 1:
            self.item_count = 1

        if abs(zipfian_constant - DEFAULT_ZIPFIAN_CONSTANT) < 1e-9:
            self.generator = ZipfianGenerator(
                0, DEFAULT_ITEM_COUNT - 1, DEFAULT_ZIPFIAN_CONSTANT, zetan=ZETAN_10B
            )
        else:
            self.generator = ZipfianGenerator(
                0, DEFAULT_ITEM_COUNT - 1, zipfian_constant
            )

    @staticmethod
    def fnv_hash_64(val: int) -> int:
        """
        Performs 64-bit FNV-1a non-cryptographic hashing matching upstream YCSB and Java/Go/Rust implementations.
        Converts negative signed 64-bit integers to positive via two's complement absolute value.
        """
        h = FNV_OFFSET_BASIS_64
        for i in range(8):
            byte = (val >> (i * 8)) & 0xFF
            h = ((h ^ byte) * FNV_PRIME_64) & 0xFFFFFFFFFFFFFFFF

        signed_h = h if h < 0x8000000000000000 else h - 0x10000000000000000
        if signed_h < 0:
            return 0 if signed_h == -0x8000000000000000 else -signed_h
        return signed_h

    def next_value(self) -> int:
        """Generates the next scrambled Zipfian integer in [min_val, max_val]."""
        raw = self.generator.next_value()
        scrambled = ScrambledZipfianGenerator.fnv_hash_64(raw) % self.item_count
        return self.min_val + scrambled


class SkewedLatestGenerator:
    """
    Generates keys with a Zipfian skew towards the most recently inserted records.
    Used in YCSB Workload D ("Read Latest").
    Caches zeta and eta parameters across reads since the basis sequence only increments on insert operations.
    """

    class _ZipfianParams:
        def __init__(
            self, item_count: int, zipfian_constant: float, zeta2_theta: float
        ):
            self.item_count = item_count
            self.zetan = ZipfianGenerator.compute_zeta(item_count, zipfian_constant)
            denom = 1.0 - zeta2_theta / self.zetan
            if abs(denom) < 1e-12:
                self.eta = 1.0
            else:
                self.eta = (
                    1.0 - math.pow(2.0 / float(item_count), 1.0 - zipfian_constant)
                ) / denom

    def __init__(
        self,
        basis: Union[int, Callable[[], int], Any],
        zipfian_constant: float = DEFAULT_ZIPFIAN_CONSTANT,
    ):
        self.basis = basis
        self.zipfian_constant = zipfian_constant
        self.zeta2_theta = 1.0 + math.pow(0.5, zipfian_constant)
        self.alpha = 1.0 / (1.0 - zipfian_constant)
        initial_max = self._get_max()
        self._cached_params = self._ZipfianParams(
            max(2, initial_max), zipfian_constant, self.zeta2_theta
        )

    def _get_max(self) -> int:
        if callable(self.basis):
            return self.basis()
        elif hasattr(self.basis, "value"):
            return self.basis.value
        elif hasattr(self.basis, "get"):
            return self.basis.get()
        return int(self.basis)

    def next_value(self) -> int:
        """Generates the next skewed latest integer."""
        max_item = self._get_max()
        if max_item <= 1:
            return 0

        params = self._cached_params
        if params is None or params.item_count != max_item:
            params = self._ZipfianParams(
                max_item, self.zipfian_constant, self.zeta2_theta
            )
            self._cached_params = params

        u = random.random()
        uz = u * params.zetan

        if uz < 1.0:
            return max_item - 1
        if uz < 1.0 + math.pow(0.5, self.zipfian_constant):
            return max(0, max_item - 2)

        offset = int(
            float(max_item) * math.pow(params.eta * u - params.eta + 1.0, self.alpha)
        )
        key = max_item - 1 - offset
        return max(0, key)


# Module-level aliases for clean backwards compatibility
compute_zeta = ZipfianGenerator.compute_zeta
fnv_hash_64 = ScrambledZipfianGenerator.fnv_hash_64

File: node/src/benchmarks/ycsb/generator.ts
export interface NumberGenerator {
  nextLong(): bigint;
  nextInt(): number;
}

export const DEFAULT_ZIPFIAN_CONSTANT = 0.99;
export const DEFAULT_ITEM_COUNT = 10000000000n;
export const ZETAN_1K = 7.728953217284738;
export const ZETAN_10B = 26.46902820178302;
export const FNV_OFFSET_BASIS_64 = 0xcbf29ce484222325n;
export const FNV_PRIME_64 = 0x100000001b3n;

/**
 * Generates random integers following a Zipfian distribution over [base, base + items - 1].
 */
export class ZipfianGenerator implements NumberGenerator {
  private items: bigint;
  private base: bigint;
  private zipfianConstant: number;
  private alpha: number;
  private zetan: number;
  private eta: number;
  private zeta2theta: number;

  constructor(
    min: number | bigint,
    max: number | bigint,
    zipfianConstant: number = DEFAULT_ZIPFIAN_CONSTANT,
    zetan?: number,
  ) {
    const minBig = BigInt(min);
    const maxBig = BigInt(max);
    let items = maxBig - minBig + 1n;
    if (items < 1n) {
      items = 1n;
    }

    this.items = items;
    this.base = minBig;
    this.zipfianConstant = zipfianConstant;
    this.zeta2theta = 1.0 + Math.pow(0.5, zipfianConstant);
    this.alpha = 1.0 / (1.0 - zipfianConstant);
    this.zetan = zetan ?? computeZeta(items, zipfianConstant);
    this.eta =
      (1.0 - Math.pow(2.0 / Number(items), 1.0 - zipfianConstant)) /
      (1.0 - this.zeta2theta / this.zetan);
  }

  public nextLong(): bigint {
    if (this.items <= 1n) {
      return this.base;
    }

    const u = Math.random();
    const uz = u * this.zetan;

    if (uz < 1.0) {
      return this.base;
    }

    if (uz < this.zeta2theta) {
      return this.base + 1n;
    }

    const offset = BigInt(
      Math.floor(
        Number(this.items) *
          Math.pow(this.eta * u - this.eta + 1.0, this.alpha),
      ),
    );
    const result = this.base + offset;
    const maxVal = this.base + this.items - 1n;
    if (result > maxVal) {
      return maxVal;
    }
    return result;
  }

  public nextInt(): number {
    return Number(this.nextLong());
  }
}

/**
 * Generates a scrambled Zipfian distribution scattering popular items across the key space.
 */
export class ScrambledZipfianGenerator implements NumberGenerator {
  private generator: ZipfianGenerator;
  private min: bigint;
  private itemCount: bigint;

  constructor(
    min: number | bigint,
    max: number | bigint,
    zipfianConstant: number = DEFAULT_ZIPFIAN_CONSTANT,
  ) {
    const minBig = BigInt(min);
    const maxBig = BigInt(max);
    this.min = minBig;
    this.itemCount = maxBig - minBig + 1n;

    if (Math.abs(zipfianConstant - DEFAULT_ZIPFIAN_CONSTANT) < 1e-9) {
      this.generator = new ZipfianGenerator(
        0n,
        DEFAULT_ITEM_COUNT - 1n,
        DEFAULT_ZIPFIAN_CONSTANT,
        ZETAN_10B,
      );
    } else {
      this.generator = new ZipfianGenerator(
        0n,
        DEFAULT_ITEM_COUNT - 1n,
        zipfianConstant,
      );
    }
  }

  public nextLong(): bigint {
    const rawZipfian = this.generator.nextLong();
    const hashed = fnvHash64(rawZipfian);
    return this.min + (hashed % this.itemCount);
  }

  public nextInt(): number {
    return Number(this.nextLong());
  }
}

/**
 * Generates uniform random integers in [min, max].
 */
export class UniformIntegerGenerator implements NumberGenerator {
  private min: bigint;
  private range: bigint;

  constructor(min: number | bigint, max: number | bigint) {
    this.min = BigInt(min);
    this.range = BigInt(max) - BigInt(min) + 1n;
  }

  public nextLong(): bigint {
    if (this.range <= 1n) {
      return this.min;
    }
    const randFraction = Math.random();
    const offset = BigInt(Math.floor(randFraction * Number(this.range)));
    return this.min + offset;
  }

  public nextInt(): number {
    return Number(this.nextLong());
  }
}

/**
 * Generates keys with a Zipfian skew towards the most recently inserted records.
 */
export class SkewedLatestGenerator implements NumberGenerator {
  private basis: {value: bigint};
  private zipfianConstant: number;
  private zeta2theta: number;
  private alpha: number;
  private cachedParams: ZipfianParams;

  constructor(
    basis: {value: bigint},
    zipfianConstant: number = DEFAULT_ZIPFIAN_CONSTANT,
  ) {
    this.basis = basis;
    this.zipfianConstant = zipfianConstant;
    this.zeta2theta = 1.0 + Math.pow(0.5, zipfianConstant);
    this.alpha = 1.0 / (1.0 - zipfianConstant);

    const initial = basis.value < 2n ? 2n : basis.value;
    const zetan = computeZeta(initial, zipfianConstant);
    const eta =
      (1.0 - Math.pow(2.0 / Number(initial), 1.0 - zipfianConstant)) /
      (1.0 - this.zeta2theta / zetan);

    this.cachedParams = {
      itemCount: initial,
      zetan,
      eta,
    };
  }

  public nextLong(): bigint {
    const maxBasis = this.basis.value;
    if (maxBasis <= 1n) {
      return 0n;
    }

    if (this.cachedParams.itemCount !== maxBasis) {
      const zetan = computeZeta(maxBasis, this.zipfianConstant);
      const eta =
        (1.0 - Math.pow(2.0 / Number(maxBasis), 1.0 - this.zipfianConstant)) /
        (1.0 - this.zeta2theta / zetan);
      this.cachedParams = {
        itemCount: maxBasis,
        zetan,
        eta,
      };
    }

    const {zetan, eta} = this.cachedParams;
    const u = Math.random();
    const uz = u * zetan;

    if (uz < 1.0) {
      return maxBasis - 1n;
    }
    if (uz < this.zeta2theta) {
      const res = maxBasis - 2n;
      return res < 0n ? 0n : res;
    }

    const offset = BigInt(
      Math.floor(Number(maxBasis) * Math.pow(eta * u - eta + 1.0, this.alpha)),
    );
    const key = maxBasis - 1n - offset;
    return key < 0n ? 0n : key;
  }

  public nextInt(): number {
    return Number(this.nextLong());
  }
}

/**
 * Computes or approximates the generalized harmonic number zeta(n, theta) in O(1) time
 * using precomputed constants and Euler-Maclaurin integral approximation (Jim Gray et al., SIGMOD '94).
 */
export function computeZeta(n: number | bigint, theta: number): number {
  const num = Number(n);
  if (Math.abs(theta - DEFAULT_ZIPFIAN_CONSTANT) < 1e-9) {
    if (num >= 1000) {
      return (
        ZETAN_1K +
        (Math.pow(num, 1.0 - theta) - Math.pow(1000.0, 1.0 - theta)) /
          (1.0 - theta)
      );
    }
    return exactZeta(num, theta);
  }

  const n0 = Math.min(num, 1000);
  let sum = exactZeta(n0, theta);
  if (num > n0) {
    sum +=
      (Math.pow(num, 1.0 - theta) - Math.pow(n0, 1.0 - theta)) / (1.0 - theta);
  }
  return sum;
}

/**
 * Computes the 64-bit FNV-1a hash matching standard YCSB key scrambling.
 * Uses two's complement absolute value to guarantee non-negative signed 64-bit integer parity.
 */
export function fnvHash64(value: number | bigint): bigint {
  let hash = FNV_OFFSET_BASIS_64;
  let v = BigInt.asUintN(64, BigInt(value));
  for (let i = 0; i < 8; i++) {
    const octet = v & 0xffn;
    v >>= 8n;
    hash ^= octet;
    hash = BigInt.asUintN(64, hash * FNV_PRIME_64);
  }
  const signedHash = BigInt.asIntN(64, hash);
  if (signedHash < 0n) {
    if (signedHash === -0x8000000000000000n) {
      return 0n;
    }
    return -signedHash;
  }
  return signedHash;
}

interface ZipfianParams {
  itemCount: bigint;
  zetan: number;
  eta: number;
}

function exactZeta(n: number, theta: number): number {
  let sum = 0.0;
  for (let i = 0; i < n; i++) {
    sum += 1.0 / Math.pow(i + 1, theta);
  }
  return sum;
}
